keps/0007-command-asserts.md

```diff
@@ -0,0 +1,103 @@
+---
+kep-number: 7
+short-desc: Assertions for CLI commands
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2020-10-28
+last-updated: 2020-10-28
+status: provisional
+---
+
+# Assertions for CLI commands
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [User Story 1](#user-story-1)
+    - [User Story 2](#user-story-2)
+  - [Implementation Details](#implementation-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+By asserting CLI commands, kuttl can support many additional test scenarios that examing the internal state of an operator.
+
+## Motivation
+
+Operators and custom resources can be used to deploy complex applications. For example, KUDO provides operators for Apache Cassandra and Apache Kafka. Deploying custom resources of these operators may not always result in changed Kubernetes resources but also in changes that aren't surfaced by Kubernetes resources but are internal to the operator. Asserting these changes only works through accessing service APIs or running commands and parsing their output. kuttl should provide tools to assert such output.
+
+### Goals
+
+- Assert that commands examining the internal state of an operator succeed or fail
+- Keep eventual consistency when asserting commands
+
+### Non-Goals
+
+- Introduce "flows" to determine when to run commands. I.e. to run commands once before or after asserting resources
+
+## Proposal
+
+Introduce commands as part of `TestAssert`. These commands are run at a regular interval like resource assertions until they either succeed or a timeout of the assert is reached. If any of the commands fail, the assert is considered to have failed. A command's return value determines if a command failed or not.
+
+As a start, complex commands that parse output can be described using shell environments. At a later stage we can implement additional tools to simplify common tasks. E.g. to search with a regular expression in command output.
+
+### User Stories
+
+#### User Story 1
+
+The JVM options of a KUDO operator can be tuned using parameters. On deployment, the JVM options are logged in the pods running the application provided by the operator. To assert that a parameter change resulted in changed JVM options we need to run `kubectl logs pod-name` and search the logs for a specific string once the operator has been deployed.
+
+#### User Story 2
+
+An operator can be configured to encrypt it's API. To assert that the API endpoints are indeed encrypted we can run a specific `curl` command and check that it returns successfully once the operator has been deployed.
+
```

Comment:
it might be interesting to parse the test logs... I'm not sure how that would be done.. but we have the ability to have background processes which are used for in development operators... you may want to parse the controller output which is in the test stdout.

If not... we should add this to the list of future considerations at the bottom of this kep